Samples/Adventure/AdventureSocketClient/MudServerHandler.cs

```diff
@@ -0,0 +1,165 @@
+            {
+                string msg = (string)message;
+                // Generate and write a response.
+                this.ProcessMessage(context, msg);
+            }
+            catch (Exception ex)
+            {
+
+                Console.WriteLine("Exception in ChannelRead0: " + ex);
+            }
+        }
+
+        public override void ChannelInactive(IChannelHandlerContext context)
+        {
+            killPlayer();
+        }
+
+        private async void killPlayer()
+        {
+            if (_player != null)
+            {
+                await _player.Play("End"); // disconnect
+                await _player.UnSubscribe(_messageInterface); // Unsubscribe - Observers pattern
+            }
+
+            await GrainClient.GrainFactory.DeleteObjectReference<IMessage>(_messageInterface);
+
+            _message = null;
+            _messageInterface = null;
+            _player = null;
+
+        }
+
+
+
+        private async void ProcessMessage(IChannelHandlerContext contex, string msg)
+        {
+            string response = "";
+            bool close = false;
+
+            try
+            {
+
+                if (string.IsNullOrEmpty(msg))
+                {
+                    response = "Please type something.\r\n";
+                }
+                else if (waitingForName)
+                {
+                    waitingForName = false;
+
+                    // Create player
+                    Guid playerGuid = Guid.NewGuid();
+
+                    _player = GrainClient.GrainFactory.GetGrain<IPlayerGrain>(playerGuid);
+                    await _player.SetName(msg);
+
+
+                    // Implement Observers pattern - https://dotnet.github.io/orleans/Documentation/Getting-Started-With-Orleans/Observers.html
+                    _message = new Message(contex);
+
+                    //Create a reference for Message usable for subscribing to the observable grain.
+                    _messageInterface = await GrainClient.GrainFactory.CreateObjectReference<IMessage>(_message);
+                    await _player.Subscribe(_messageInterface);
+
+                    // Put the player into the default Room
+                    var room1 = GrainClient.GrainFactory.GetGrain<IRoomGrain>(0);
+                    await _player.SetRoomGrain(room1);
+
+                    response = await _player.Play("look");
+
+                }
+                else if (string.Equals("End", msg, StringComparison.OrdinalIgnoreCase))
+                {
+                    response = "Have a good day!\r\n";
+
+                    close = true;
+                }
+                else
+                {
+                    response = await _player.Play(msg);
+                }
+            }
+            catch (Exception ex)
+            {
+                response = "Unexpected exception: " + ex;
+                close = true;
+            }
+
+            Task wait_close = contex.WriteAndFlushAsync(response + "\r\n");
```

Comment:
Not really as WriteAndFlushAsync() must be executed even if "close" is false.

But I have refactored the code to Await the write() function, and if "close" is true I do Await on the CloseAsync.

await context.WriteAndFlushAsync(response + "\r\n");
if (close)
{                
    await context.CloseAsync();
}